contracts/tokenWhitelist.sol

```diff
@@ -152,7 +177,91 @@ contract TokenWhitelist is ENSResolvable, Controllable, Transferrable {
+        }
+    }
+
+    /// @notice Remove method signatures(methodIDs) from the whitelist.
+    /// @param _methods method signatures to be removed.
+    function removeMethodIds(bytes4[] calldata _methods) external onlyAdmin {
+        for (uint i = 0; i < _methods.length; i++) {
+            bytes4 method = _methods[i];
+            if (_methodIdWhitelist[method]) {
+                _methodIdWhitelist[method] = false;
+                emit RemovedMethodId(method);
+            }
+        }
+    }
+
+    /// @notice Add exclusive methods supported for a specific token that are not in the general whitelist.
+    /// @param _token token address.
+    /// @param _methods method signatures to be added.
+    function addExclusiveMethods(address _token, bytes4[] calldata _methods) external onlyAdmin {
+        for (uint i = 0; i < _methods.length; i++) {
+            bytes4 method = _methods[i];
+            require(_tokenInfoMap[_token].available, "non-existing token");
+            if (!(_methodIdWhitelist[method] || _tokenInfoMap[_token].exclusiveMethods[method])) {
+                _tokenInfoMap[_token].exclusiveMethods[method] = true;
+                emit AddedExclusiveMethod(_token, method);
+            }
+        }
+    }
+
+    /// @notice Remove exclusive methods supported for a specific token.
+    /// @param _token token address.
+    /// @param _methods method signatures to be removed.
+    function removeExclusiveMethods(address _token, bytes4[] calldata _methods) external onlyAdmin {
+        for (uint i = 0; i < _methods.length; i++) {
+            bytes4 method = _methods[i];
+            require(_tokenInfoMap[_token].available, "non-existing token");
+            if (_tokenInfoMap[_token].exclusiveMethods[method]){
+                _tokenInfoMap[_token].exclusiveMethods[method] = false;
+                emit RemovedExclusiveMethod(_token, method);
+            }
+        }
+    }
+
+    /// @notice based on the method it returns the recipient address and amount/value, ERC20 specific.
+    /// @param _data is the transaction payload.
+    function getERC20RecipientAndAmount(address  _destination, bytes calldata _data) external view returns (address, uint) {
+        // Require that there exist enough bytes for encoding at least a method signature + data in the transaction payload:
+        // 4 (signature)  + 32(address or uint256)
+        require(_data.length >= 4 + 32, "not enough method-encoding bytes");
+        // Get the method signature
+        bytes4 signature = _data._bytesToBytes4(0);
+        // Check if method Id is supported
+        require(isERC20MethodSupported(_destination, signature), "unsupported method");
+        //to is the recipient's address and amount is the value to be transferred
+        address to;
+        uint amount;
+        if (signature == _BURN){
+            // 4 (signature) + 32(uint256)
+            require(_data.length >= 4 + 32, "not enough data for burn");
+            to = _destination;
```

Comment:
how about returning here? its ridiculous to set variables and then have one return ... return here would make the code more readable.